Adds Network::from_core_arg and links it to Network::from_str

sr-gi opened the pull request:
Comming from https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/247

`Network::from_str` only considers `rust-bitcoin` string as possible inputs to create a `Network` instance. This PR adds a new method to `Network`, `from_core_arg`, which is complementary to the existing `Network::to_core_arg`. This method allows the conversion between `bitcoind -network` string and `Network` variants.

This also links `Network::from_str` to `Network::from_core_arg` so the default case on the former calls the latter, and an error is only returned if none of the cases match.

sr-gi commented:
@apoelstra I've taken the liberty of adding `mainnet` and `testnet` to `from_core_arg` even though they are not strictly returned by `bitcoind`. I can remove those if needed.

Also, if you feel the link between `from_core_arg` and `from_str` is unnecessary (or not something you'd like to have a the moment) I'm happy to revert that one commit. Otherwise, they may be squashable.

apoelstra commented:
Thanks!

I think that we should only support `main` (not `mainnet`) and similar for `test`. Ideally the two `_core_arg` methods are exact inverses of each other.

As for `from_str`, I'm happy either way. Personally I think that `Network::from_str` does not have a clear purpose or semantics, so I don't know what to do with it :).

tcharding commented on bitcoin/src/network/constants.rs:
```diff
@@ -144,7 +166,7 @@ impl FromStr for Network {
-            _ => return Err(ParseNetworkError(s.to_owned())),
+            _ => return Network::from_core_arg(s),
         };
         Ok(network)
```
Perhaps this would be cleaner
```rust
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        // Some comment about why "bitcoin" is used as well as "mainnet".
        if s == "bitcoin" {
            Ok(Network::Bitcoin)
        } else {
            Network::from_core_arg(s)
        }
    }
```

sr-gi replied:
Agreed modulo `testnet` is agreed upon as an option for `from_core_args`

tcharding commented on bitcoin/src/network/constants.rs:
```diff
@@ -119,6 +119,28 @@ impl Network {
+        use Network::*;
+
+        let network = match core_arg {
+           "main" | "mainnet" => Bitcoin,
+           "test" | "testnet" => Testnet,
```
Were "mainnet" and "testnet" used in old versions of bitcoind? 22 and 23 use "main" and "test", I don't have any older versions laying around.

Kixunil replied:
AFAIR `mainnet` was never used, not sure about `testnet`. I'd rather not accept strings that aren't used in Core.

Also does anyone remember if Core is case-sensitive? My RAM is full RN so can't spawn a VM.

sr-gi replied:
I don't think they were ever used, we may not want to have this in `from_core_arg` but we may for `from_str`.

sanket1729 replied:
There are two distinct ways of using bitcoind/bitcoin-cli. As Sr-gi mentions, 

1) As an named option `chain=` `bitcoind -chain={main, test, regtest, signet}`
2) Directly as an unnamed option `bitcoind --testnet` / `bitcoind --signet`, / `bitcoind --regtest`

Given this, I agree with the current implementation of this PR.

sanket1729 replied:
For cross refernce: https://github.com/bitcoin/bitcoin/blob/cac29f5cd66f005ce65edd697990d495a904a1f1/src/chainparamsbase.cpp#L18-L29

Kixunil replied:
I don't think that is a good reason since the documentation explicitly mentions `-chain` argument in both `to_core_arg` and `from_core_arg`. I don't really like accepting `-` variants since they can't represent mainnet and we would have to require `-`. For something exotic such as Rust replacement of `bitcoin-cli` people should just write their own I think.

Perhaps we should rename the functions to `{to,from}_core_chain_arg` to be entirely clear but could be annoying.

tcharding commented on bitcoin/src/network/constants.rs:
````diff
@@ -119,6 +119,28 @@ impl Network {
+
+    /// Converts a `bitcoind -chain` argument name to its equivalent `Network`.
+    ///
+    /// ```bash,no_run
````
We can remove the `,no_run` - that was a mistake on `to_core_arg` by me that I now see I haven't fixed yet. Will fix now.

tcharding replied:
#1430

sr-gi replied:
Addressed in [cb5dea4](https://github.com/rust-bitcoin/rust-bitcoin/pull/1428/commits/cb5dea45026e59c8948f149ce46af389c09abd31)

Kixunil commented on bitcoin/src/network/constants.rs:
```diff
@@ -144,7 +166,7 @@ impl FromStr for Network {
             "signet" => Signet,
             "regtest" => Regtest,
-            _ => return Err(ParseNetworkError(s.to_owned())),
+            _ => return Network::from_core_arg(s),
```
I'm not really convinced this is a good idea. It's basically [robustness principle which is harmful](https://www.ietf.org/archive/id/draft-iab-protocol-maintenance-05.html)

(Interestingly the frequency of attempts at robustness principle I replied to is unusually high last few months...)

tcharding replied:
However we decide to do it, accepting "mainnet" seems like a win to me, bitcoins overloading of "bitcoin" for the network is annoying when we have a term used in the common vernacular that better describes it.

Kixunil replied:
Yeah, I'm open to that one. I'm even tempted to drop `bitcoin` but I think it was already changed in the past and this would break user-facing code if blindly upgraded. E.g. electrs uses this in argument/configuration parsing.

sr-gi replied:
This is mainly why I went for keeping both instead of dropping `bitcoin`. I'm happy with any of the options: either accepting what is used in common vernacular (main/mainnet, test/testnet) or going for only what Core returns (main/test).

Kixunil reviewed (commented):
Concept ACK

sr-gi commented:
It looks like there's a bit of discrepancy with whether or not to accept `mainnet` and `testnet` as well as on whether `from_str` should be linked to `from_core_args`.

I personally don't have any real preference, just having a simple `from_core_args` will already simplify `Network` parsing downstream for me. On the other hand, it may be worth agreeing on what the functionality of `from_str` should be.

Kixunil commented:
In that case I suggest focusing on `from_core_arg` first and leave `from_str` for a separate discussion.

For `from_core_arg` I'd want it to work 100% exactly like Core: whichever argument Core accepts we should accept and whichever it rejects we should reject.

But since your use case is RPC I'm not even sure if we should assume that Core will always use the same values as for command line arguments. Feels like we should have a separate function (even if the code is currently identical).

sr-gi commented:
> In that case I suggest focusing on `from_core_arg` first and leave `from_str` for a separate discussion.
> 
> For `from_core_arg` I'd want it to work 100% exactly like Core: whichever argument Core accepts we should accept and whichever it rejects we should reject.

I've just checked Core for this. `getblockchaininfo` states: 

```
"chain" : "str",    (string) current network name (main, test, signet, regtest)
```

Whereas `bitcoind help` for `-chain` states:

```
-chain=<chain>
     Use the chain <chain> (default: main). Allowed values: main, test,
     signet, regtest
```

And combinations with different case sensitivity (as well as wrong arguments) fail:

```
> bitcoind -chain=Main

************************
EXCEPTION: St13runtime_error
Invalid combination of -regtest, -signet, -testnet and -chain. Can use at most one.
bitcoin in AppInit()

```

That does not look like the proper way of handling an unknown argument by Core. by the way, it may be worth reporting.

> But since your use case is RPC I'm not even sure if we should assume that Core will always use the same values as for command line arguments. Feels like we should have a separate function (even if the code is currently identical).

Not sure I follow here though.

Kixunil commented:
> Not sure I follow here though.

Hypothetically in the future there could be a new network that is named differently in CLI arg and in RPC. Then the code would break. But now that I think about it, it seems reasonably unlikely.

apoelstra commented:
> Hypothetically in the future there could be a new network that is named differently in CLI arg and in RPC. Then the code would break. But now that I think about it, it seems reasonably unlikely.

Yeah, they use literally the same code paths in Core. I think it's safe to assume they won't break this.

apoelstra commented:
cb5dea45026e59c8948f149ce46af389c09abd31 looks good to me (the first commit). I'd rather we not accept `testnet` or `mainnet` but I don't feel strongly.

The second commit, which changes `from_str`, I think we should drop, at least for now.

sanket1729 reviewed (commented):
I don't really feel so strongly about the second commit. But can ACK this if others have no complaints

sr-gi commented:
Given the current feedback, I think all of us have agreed to accept cb5dea45026e59c8948f149ce46af389c09abd31 (some of you modulo dropping the `mainnet` and `testnet` variants) and drop a5d8987092d85260ad2f88d5efe4abd6a9973b82, at least for now.

If so, I'll do so and rework https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/247 based on the state of this PR.

sr-gi commented:
Dropped https://github.com/rust-bitcoin/rust-bitcoin/commit/a5d8987092d85260ad2f88d5efe4abd6a9973b82 and refactored https://github.com/rust-bitcoin/rust-bitcoin/commit/cb5dea45026e59c8948f149ce46af389c09abd31 to remove `mainnet` and `testnet`.

Kixunil reviewed (approved):
ACK bd1eb29f61f451885d1f11b05f61023493584fcb

A test that checks `to_core_arg` roundtrips would be nice. (Can be a separate commit.)

sr-gi commented:
> ACK [bd1eb29](https://github.com/rust-bitcoin/rust-bitcoin/commit/bd1eb29f61f451885d1f11b05f61023493584fcb)
> 
> A test that checks `to_core_arg` roundtrips would be nice. (Can be a separate commit.)

Added in [d7006ef](https://github.com/rust-bitcoin/rust-bitcoin/pull/1428/commits/d7006ef80d1562da635f10cc435a7cb45f3e440a)

Kixunil commented on bitcoin/src/network/constants.rs:
```diff
@@ -599,4 +599,19 @@ mod tests {
+
+        for (net, core_arg) in &expected_pairs {
+            assert_eq!(Network::from_core_arg(core_arg), Ok(*net));
+            assert_eq!(net.to_core_arg(), *core_arg);
```
Could've been just `assert_eq!(Network::from_core_arg(net.to_core_arg()).unwrap(), net);` but OK.

Kixunil replied:
Just less boilerplate :)

Kixunil reviewed (approved):
ACK d7006ef80d1562da635f10cc435a7cb45f3e440a

apoelstra reviewed (approved):
ACK d7006ef80d1562da635f10cc435a7cb45f3e440a

sr-gi commented:
@apoelstra the merge commit contains:

"This also links `Network::from_str` to `Network::from_core_arg` so the default case on the former calls the latter, and an error is only returned if none of the cases match." which is not true anymore 😅

apoelstra commented:
@sr-gi oops! I will make a mental note to re-read the PR description of things before signing off on them.

sr-gi commented:
> @sr-gi oops! I will make a mental note to re-read the PR description of things before signing off on them.

My bad tbf, I completely forgot to edit that out after rebasing 😅
